cores/esp32/IPAddress.cpp

```diff
@@ -364,5 +244,168 @@ bool IPAddress::fromString6(const char *address) {
+}
+
+bool IPAddress::operator==(const uint8_t* addr) const
+{
+    // IPv4 only comparison to byte pointer
+    // Can't support IPv6 as we know our type, but not the length of the pointer
+    return _type == IPv4 && memcmp(addr, &_address.bytes[IPADDRESS_V4_BYTES_INDEX], sizeof(uint32_t)) == 0;
+}
+
+uint8_t IPAddress::operator[](int index) const {
+    if (_type == IPv4) {
+        return _address.bytes[IPADDRESS_V4_BYTES_INDEX + index];
+    }
+    return _address.bytes[index];
+}
+
+uint8_t& IPAddress::operator[](int index) {
+    if (_type == IPv4) {
+        return _address.bytes[IPADDRESS_V4_BYTES_INDEX + index];
+    }
+    return _address.bytes[index];
+}
+
+size_t IPAddress::printTo(Print& p) const
+{
+    size_t n = 0;
+
+    if (_type == IPv6) {
+        // IPv6 IETF canonical format: compress left-most longest run of two or more zero fields, lower case
+        int8_t longest_start = -1;
+        int8_t longest_length = 1;
+        int8_t current_start = -1;
+        int8_t current_length = 0;
+        for (int8_t f = 0; f < 8; f++) {
+            if (_address.bytes[f * 2] == 0 && _address.bytes[f * 2 + 1] == 0) {
+                if (current_start == -1) {
+                    current_start = f;
+                    current_length = 1;
+                } else {
+                    current_length++;
+                }
+                if (current_length > longest_length) {
+                    longest_start = current_start;
+                    longest_length = current_length;
+                }
+            } else {
+                current_start = -1;
+            }
+        }
+        for (int f = 0; f < 8; f++) {
+            if (f < longest_start || f >= longest_start + longest_length) {
+                uint8_t c1 = _address.bytes[f * 2] >> 4;
+                uint8_t c2 = _address.bytes[f * 2] & 0xf;
+                uint8_t c3 = _address.bytes[f * 2 + 1] >> 4;
+                uint8_t c4 = _address.bytes[f * 2 + 1] & 0xf;
+                if (c1 > 0) {
+                    n += p.print((char)(c1 < 10 ? '0' + c1 : 'a' + c1 - 10));
+                }
+                if (c1 > 0 || c2 > 0) {
+                    n += p.print((char)(c2 < 10 ? '0' + c2 : 'a' + c2 - 10));
+                }
+                if (c1 > 0 || c2 > 0 || c3 > 0) {
+                    n += p.print((char)(c3 < 10 ? '0' + c3 : 'a' + c3 - 10));
+                }
+                n += p.print((char)(c4 < 10 ? '0' + c4 : 'a' + c4 - 10));
+                if (f < 7) {
+                    n += p.print(':');
+                }
+            } else if (f == longest_start) {
+                if (longest_start == 0) {
+                    n += p.print(':');
+                }
+                n += p.print(':');
+            }
+        }
+        // add a zone if zone-id is non-zero
+        if(_zone > 0){
+            n += p.print('%');
+            char if_name[NETIF_NAMESIZE];
+            netif_index_to_name(_zone, if_name);
```

Comment:
Any particular reason why you are using the (optional) string name format, e.g. "%eth0", rather than just outputting the numeric zone identifier "%2". In RFC 4007 the numeric format is considered the canonical "implementation SHOULD support at least numerical index", whereas only "MAY support other kinds".

Reply:
This is a question that should be asked in ESP-IDF's repo. We only consume what is used there. If our network stack returns netif names, this is what you will always get. To have numerical value, you would have to enter it manually (wrongly for that system). That code is not meant to be portable

Reply:
The toString() and fromString() are just string representations, with the zone. So long as toString(), e.g. displayed to user, is parseable by fromString(), then it should be irrelevant if you default to the zone_id number (2, 3, etc), and some other app parses it to and from an interface name.

Are you ever going to swap strings between the two programs? Even then the fromString() should handle both/either input, as either is valid according to RFC specs. This could easily be checked by trying to both parse as a number (atoi()?) and then try parse as an interface name (or the other way around).

But, for a first merge, only supporting one (only interface names) should be fine.

Reply:
what does the RFC say about zone names?

Reply:
The RFC has gaps ... the main RFC 4007 doesn't put any limits on non-number zone names. It doesn't even limit the numeric except to say a positive number. The semantics of the default interface are also not well defined. Even the way % is used as a separator is not clear.

This caused all sorts of problems. e.g. not being able to use something in a browser like "http://[fe80::1111.2222.3333.4444]/ as a fixed web address for a device (e.g. in a QR Code), i.e. relying on default zone. Support differs between browser and OS, and a lot of the browsers stopped supported because there wasn't a consensus on whether it was % or %25 in a URL.

Zone names are all over the place. Different tools do it differently (ping, curl, etc) and interfaces can be anything,

e.g. On my linux machine I have `ping fe80::c256:27ff:fe73:4b13%enx387c76021431` (the same as `%2`). 

Whether they are case sensitive is also not specified.

I think some general recommendations are to stick with lower case, short, and alphanumerics.

It looks like LWIP uses 2 characters + a number (mapping to an underlying numeric byte zone ID).

Reply:
> what does the RFC say about zone names?

Yesterday I found this article: https://labs.ripe.net/author/philip_homburg/whats-the-deal-with-ipv6-link-local-addresses/

Reply:
what I figured out is that the zone name can be anything that can identify the interface locally. We can do basically the same and use `netif->desc` as zone. That would give the lower case name of the interface as it's defined in our stack. We will go with `%` as delimiter and be done. Agreed?

Reply:
Agree to me. I initially append the number of the zone like `%st1` or `%en2`. But in this case there can't be 2 WIFI or 2 ETH, so the number is not needed.

So we're going for `%st` and `%en`?

Reply:
we can have multiple ethernets :) it would be as defined in IDF/LwIP and would be searchable, meaning you can get the exact netif from the zone.

Reply:
`toString()`

Can remain as is. 

If _zone == 0, then it does not need to be printed, e.g. fe80::1 with _zone = 0 is just "fe80::1"

Otherwise, we are using the optional implementation in RFC 4007 to use another identifier, in this case following the recommendation using the interface name, via `netif_index_to_name()`

`netif.c` implements `netif_index_to_name()` as the first two characters of the name, plus the (index - 1). See https://github.com/particle-iot/lwip/blob/832490eec8ca35bad435fec104512cfbdaf994bf/src/core/netif.c#L1687

This satisfies the criteria of being unique (because index is unique), plus being somewhat human readable.

e.g. on my laptop where interface/zone %2 name is enx387c76021431, then it would become "fe80::1%en1" (note that the index of 2 becomes a name of 1)

I think an assumption that there is only one interface of each type is short sighted, so don't think '%st' is a good idea.

Note that LwIP doesn't append the zone, but (zone - 1), i.e. "%st0" for zone 1, "%en1" for zone 2.

Any is fine, so long as fromString() handles the reverse, although keeping the code as is (the LwIP `netif_index_to_name()`) may be best.


`fromString()`

Needs changes to at least support numeric indexes passed in.

RFC 4007, "implementation SHOULD support at least numerical indices" (p. 16), so fromString() should check, after the %, if the remaining address can be converted to a number, and if so set `_zone = number`. e.g. a test should parse "fe80::1%2" to _zone = 2.

"Zero SHOULD be reserved to represent the default zone", so "fe80::1%0" is also valid and should parse to _zone = 0.

"When <zone_id> is the default, the delimiter characters "%" and <zone_id> can be omitted." (p. 16), so "fe80::1" becomes _zone = 0, and without a zone index should also be the default, i.e. "fe80::1%" is also _zone = 0.

"implementation MAY support other kinds" (p. 16), with a mention that interface names would be a good candidate. i.e. if the part after the % is not a number, check if it is an interface name via `_zone = netif_name_to_index(address);`

This other kind should at least be whatever we output from `toString()`

Reply:
> `toString()` - can remain as is.

Well it would be nice to have an option to not include the zone while 'printing' it.
For example when creating a link or something the other end doesn't have anything to do with a zone as the zone is local to the device., right?

This way you don't have to copy the `IPAddress` object with a zone set to 0, just to print it.

So something like `toString(bool excludeZone = false)`

Reply:
Good with me, thanks.

Reply:
> Well it would be nice to have an option to not include the zone while 'printing' it.

Sure, I don't see any problem with an option.

> For example when creating a link or something the other end doesn't have anything to do with a zone

Both ends will require an index, just different ones, so whether you output the index on side 1 or not, you still need to change it on side 2.

More detail:

Well, for a scoped address (e.g. link local) -- one with a zone index -- it will have a zone on both ends (but different indexes). i.e. if my router uses "fe80::1" on it's LAN  interface (number 11), then for my laptop to access it I need to use the relevant zone, e.g. `ping fe80::1%3`; on another Windows machine I need to use `ping fe80::1%9`

My Windows machine reports (`ipconfig /all`):
`Link-local IPv6 Address . . . . . : fe80::4d6d:9ea:a521:d435%9(Preferred)`

For the router to ping Windows it is `ping fe80::4d6d:9ea:a521:d435%11`

Windows shows the scope index (from it's point of view) when listing it's IP addresses (where relevant, e.g. link local), although this won't match what index other clients need to use -- other clients will still need to use an index, just a different one for each.

Windows doesn't have to show the index (as it is no that useful), e.g. in contrast my laptop (`ip addr`) does not show it:
`    inet6 fe80::f949:fb86:d418:aaa4/64 scope link noprefixroute`

However it doesn't do any harm either, from my router I still need to add the index `ping fe80::f949:fb86:d418:aaa4%11`

**Default Index**: Windows actually supports default indexes (Linux does not), so from my Windows machine I can also do just `ping fe80::1` and it will work, but only because the LAN (%9) happens to be the default. But if it wasn't the default I would need to add the index. So maybe printing the IP address without an index you get lucky and it happens to work from a Windows machine.

Reply:
Yeah but those are all referencing the local zone.
My intended use case is where you let your ESP tell other devices how to reach you.
Or simply put a link on a page for a user to reach a certain node.

To give some context, part of the 'root page' of ESPEasy where it shows all nodes found via ESPEasy p2p"
![image](https://github.com/espressif/arduino-esp32/assets/3751318/3540d9af-b043-4e05-a926-bc88a5f1dd8e)

Reply:
so here is the deal:
- zone only makes sense on the ESP itself, outside of the ESP the zone is irrelevant.
- LwIP provides the zone in int, which is the LwIP netif index (not esp_netif and starts with 1, where 1 is loopback)
- Makes sense to print the zone only on the device and do not give it for outside use or discard for outside (image above)
- Global addresses have zone 0, which is not printer
- Lastly, here is how it looks currently on my ESP (I have set global and link-local IPv6)
```
ETH Got IPv6: 'eth0' (en1)
eth0: <UP,100M,FULL_DUPLEX,AUTO>
      ether 32:C6:F7:F4:2F:8C phy 0x1
      inet 192.168.254.22 netmask 255.255.255.0 broadcast 192.168.254.255
      gateway 192.168.254.1 dns 192.168.254.3
      inet6 fe80::30c6:f7ff:fef4:2f8c%en1 type LINK_LOCAL
      inet6 2a10:bac0:b000:7395:30c6:f7ff:fef4:2f8c type GLOBAL

```

Reply:
> My intended use case is where you let your ESP tell other devices how to reach you.
> Or simply put a link on a page for a user to reach a certain node.

Short answer: I think adding the option to convert without includeScope = false (with a default of true) is fine to add. It is probably a slightly nicer display (although generally the caller will still need to know their scope index to add).


Longer answer:

So @TD-er, in your example the esp32 has link-local address fe80::9af:abff:fe24:46df

If it was on my local LAN, to ping it from my Linux laptop I would have to do `ping fe80::9af:abff:fe24:46df%3`, to ping it from my Windows machine I would need to do `ping fe80::9af:abff:fe24:46df%9`

The address is fe80::9af:abff:fe24:46df, which which zone index is the LAN interface will be different for each machine.

Additional note: On Windows (Linux) I can do `ping fe80::9af:abff:fe24:46df`, because it supports default zones and the LAN (%9) happens to be the default ... but for other machines it may not be default. 

The @me-no-dev device has link-local address fe80::30c6:f7ff:fef4:2f8c (and on the device itself that happens to be zone %en1, although that is not relevant).

If it was also on my LAN, from my Linux laptop `ping fe80::30c6:f7ff:fef4:2f8c%3`, or Windows machine `ping fe80::30c6:f7ff:fef4:2f8c%9`

If by link you actually me put a URL link into a browser, you get the same issue: On Windows "http://[fe80::9af:abff:fe24:46df]" would work if the device is in the default zone.

If it is not the default zone, or you are on Linux, then it won't work in a browser at all, because none of the browsers support the '%' scope prefix. Other tools will work, e.g. ping, `curl fe80::30c6:f7ff:fef4:2f8c%3`, etc. Just not browsers (except the case of default zone on Windows).

So either @TD-er "fe80::9af:abff:fe24:46df" or @me-no-dev "fe80::30c6:f7ff:fef4:2f8c%en1" is fine to me; in both cases I can get the address and add my local index (%3, %9, etc).

Note that Linux `ip addr` lists addresses without the suffix and Windows `ipconfig` lists them with.

Without might be slightly nicer, and might work in a browser on Windows if it is the default (e.g. a single interface); and maybe one day in the future Linux will support default as well.

Reply:
Note: A quick check of Matter on my local network (a dozen or so devices) did not show any link-local published to mDNS.

All addresses were either global IPv6 (prefix 2407:8800:bc61:1340...), a local Thread ULA prefix (advertise by Google, for routing through Thread Border Routers, fdf0:de49:...), or any IPv4 (frowned upon, but allowed):

```
avahi-browse -d local _matter._tcp --resolve -t

= wlp9s0 IPv6 E3897703A3257ABC-00000000ED0B4636             _matter._tcp         local
   hostname = [48D6D57E6342.local]
   address = [2407:8800:bc61:1340:4ad6:d5ff:fe7e:6342]
   port = [5540]
   txt = []
= wlp9s0 IPv6 E3897703A3257ABC-00000000B878C0AB             _matter._tcp         local
   hostname = [B2B6343D0585560B.local]
   address = [fdf0:de49:2af5:1:887:52df:d36e:28c7]
   port = [5540]
   txt = ["T=0" "SAI=800" "SII=800"]
= wlp9s0 IPv4 e3897703a3257abc-00000000cae3c9ed             _matter._tcp         local
   hostname = [14C14ECD6D16.local]
   address = [192.168.1.218]
   port = [5540]
   txt = []
...
```

Any ordering or displaying of address should generally preference a global address (ULAs are also global) if available, e.g. @me-no-dev `2a10:bac0:b000:7395:30c6:f7ff:fef4:2f8c`, as that can be linked to.

@TD-er while a device will only have one IPv4 address, it will have multiple IPv6, rather than show the first (usually a link-local), you should show the "best", i.e. a global if available. (or better yet, list them all)